src/main/java/org/opensearch/neuralsearch/util/ProcessorDocumentUtils.java

```diff
@@ -231,6 +232,63 @@ public static Map<String, Object> unflattenJson(Map<String, Object> originalJson
         return result;
     }
 
+    /**
+     * Flattens a nested map and then flips each key/value pair.
+     * For a leaf node, the parent's path is prepended to its value.
+     * Finally, the flipped map will have the transformed value as the key,
+     * and the original flattened key as its value.
+     * e.g:
+     * map:
+     * {
+     *     "parent": {
+     *         "level1": {
+     *             "key": "value"
+     *         }
+     *     }
+     * }
+     * returns
+     * {
+     *     "parent.level1.value": "parent.level1.key"
+     * }
+     *
+     * @param map the nested map to process
+     * @return a flattened map with flipped key–value pairs
+     */
+    public static Map<String, String> flattenAndFlip(Map<String, Object> map) {
+        Map<String, String> flippedMap = new HashMap<>();
+        flattenAndFlip("", map, flippedMap);
+        return flippedMap;
+    }
+
+    /**
+     * Recursive helper method that processes the nested map that strictly contains Maps and Strings only.
+     * When a String value is encountered, the parent's path is computed
+     * and prepended to the value. The final mapping is flipped, so that
+     * the new key becomes the computed value and the new value is the flattened key.
+     * In case of duplicate keys, the values are overwritten
+     *
+     * @param path     the current key prefix (initially empty)
+     * @param map        the current map to process
+     * @param flippedMap the resulting map with flipped key–value pairs
+     */
+    private static void flattenAndFlip(String path, Map<String, Object> map, Map<String, String> flippedMap) {
+        for (Map.Entry<String, Object> entry : map.entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+            String currentKey = path.isEmpty() ? key : path + "." + key;
+            if (Objects.isNull(currentKey)) return;
```

Comment:
Shouldn't we check `key` but not `currentKey`?
```suggestion
            if (Objects.isNull(key)) return;
```

Reply:
Could you fix the unit text?